File: import_managers.py
import csv
import os
import re
from pathlib import Path

import psycopg2
from openpyxl import load_workbook

from db import getconn, putconn


def _normalize(value):
    if value is None:
        return None
    return str(value).strip()


_PHONE_RE = re.compile(r"^\+?\d[\d\s\-]{7,15}$")
_NAME_RE = re.compile(r"^[A-Za-z\s\.]{2,}$")


def _looks_like_phone(value):
    if not value:
        return False
    return bool(_PHONE_RE.match(value))


def _looks_like_name(value):
    if not value:
        return False
    return bool(_NAME_RE.match(value))


def _first_non_none(row: dict, aliases):
    lowered = {str(k).lower(): k for k in row.keys() if k is not None}
    for alias in aliases:
        if alias in lowered:
            value = _normalize(row[lowered[alias]])
            if value:
                return value
    return None


def _extract_contacts_from_row(row: dict):
    contacts = []

    rsm_name = _first_non_none(row, ["rsm name", "rsm"])
    rsm_phone = _first_non_none(row, ["ph no", "rsm\nphone number", "phone", "mobile", "mobile_no", "contact", "phone_number", "mobile_number", "sm phone number"])
    rsm_location = _first_non_none(row, ["location", "city", "branch", "branch_name", "place"])
    if rsm_name or rsm_phone:
        contacts.append({
            "name": rsm_name if _looks_like_name(rsm_name) else (rsm_phone if _looks_like_name(rsm_phone) else None),
            "phone": rsm_phone if _looks_like_phone(rsm_phone) else (rsm_name if _looks_like_phone(rsm_name) else None),
            "location": rsm_location,
            "role": "RSM",
        })

    sm_name = _first_non_none(row, ["sm name", "sm"])
    sm_phone = _first_non_none(row, ["ph no_1", "rsm\nphone number", "phone", "mobile", "mobile_no", "contact", "phone_number", "mobile_number", "sm phone number"])
    sm_location = _first_non_none(row, ["location_1", "location", "city", "branch", "branch_name", "place"])
    if sm_name or sm_phone:
        contacts.append({
            "name": sm_name if _looks_like_name(sm_name) else (sm_phone if _looks_like_name(sm_phone) else None),
            "phone": sm_phone if _looks_like_phone(sm_phone) else (sm_name if _looks_like_phone(sm_name) else None),
            "location": sm_location,
            "role": "SM",
        })

    coordinator_name = _first_non_none(row, ["coordinator", "co-ordinator", "name"])
    coordinator_phone = _first_non_none(row, ["mobile", "phone", "mobile_no", "contact", "phone_number", "mobile_number", "rsm\nphone number", "sm phone number"])
    coordinator_location = _first_non_none(row, ["location_2", "location_1", "location", "city", "branch", "branch_name", "place"])
    if coordinator_name or coordinator_phone:
        contacts.append({
            "name": coordinator_name if _looks_like_name(coordinator_name) else (coordinator_phone if _looks_like_name(coordinator_phone) else None),
            "phone": coordinator_phone if _looks_like_phone(coordinator_phone) else (coordinator_name if _looks_like_phone(coordinator_name) else None),
            "location": coordinator_location,
            "role": "Coordinator",
        })

    asm_name = _first_non_none(row, ["asm/sm/dsm name", "asm", "sm", "dsm"])
    asm_phone = _first_non_none(row, ["mobile", "phone", "mobile_no", "contact", "phone_number", "mobile_number", "rsm\nphone number", "sm phone number"])
    asm_location = _first_non_none(row, ["location_2", "location_1", "location", "city", "branch", "branch_name", "place"])
    if asm_name or asm_phone:
        contacts.append({
            "name": asm_name if _looks_like_name(asm_name) else (asm_phone if _looks_like_name(asm_phone) else None),
            "phone": asm_phone if _looks_like_phone(asm_phone) else (asm_name if _looks_like_phone(asm_name) else None),
            "location": asm_location,
            "role": "ASM/SM/DSM",
        })

    generic_name = _first_non_none(row, ["name", "manager_name", "manager", "employee_name", "contact_person"])
    generic_phone = _first_non_none(row, ["mobile", "phone", "mobile_no", "contact", "phone_number", "mobile_number", "rsm\nphone number", "sm phone number"])
    generic_location = _first_non_none(row, ["location", "city", "branch", "branch_name", "place"])
    if generic_name or generic_phone:
        contacts.append({
            "name": generic_name if _looks_like_name(generic_name) else (generic_phone if _looks_like_name(generic_phone) else None),
            "phone": generic_phone if _looks_like_phone(generic_phone) else (generic_name if _looks_like_phone(generic_name) else None),
            "location": generic_location,
            "role": None,
        })

    return contacts


def _insert_managers(conn, managers):
    cursor = conn.cursor()
    query = """
        INSERT INTO bank_managers (bank_name, name, email, phone, location, role, status)
        VALUES (%s, %s, %s, %s, %s, %s, %s)
    """
    for m in managers:
        cursor.execute(
            query,
            (
                m.get("bank_name"),
                m.get("name") or "Unknown",
                m.get("email") or f"unknown_{id(m)}@example.com",
                m.get("phone"),
                m.get("location") or "Unknown",
                m.get("role"),
                "active",
            ),
        )
    conn.commit()
    cursor.close()


def _read_csv(path: Path):
    with open(path, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        return list(reader)


def _read_excel(path: Path):
    wb = load_workbook(path, read_only=True)
    ws = wb.active
    rows = []
    for row in ws.iter_rows(min_row=1, values_only=True):
        rows.append(list(row))
    wb.close()

    header_row_idx = None
    for idx, row in enumerate(rows):
        non_none = [v for v in row if v is not None]
        if len(non_none) >= 3 and all(isinstance(v, str) for v in non_none):
            header_row_idx = idx
            break

    if header_row_idx is None:
        header_row_idx = 0

    raw_headers = rows[header_row_idx]
    headers = []
    seen = {}
    for h in raw_headers:
        if h is None:
            headers.append(None)
            continue
        key = str(h).lower()
        if key in seen:
            seen[key] += 1
            headers.append(f"{h}_{seen[key]}")
        else:
            seen[key] = 0
            headers.append(h)

    data_rows = []
    for row in rows[header_row_idx + 1:]:
        row_dict = {}
        for i, header in enumerate(headers):
            if header is not None and i < len(row):
                row_dict[header] = row[i]
        data_rows.append(row_dict)

    return data_rows


def import_managers_from_file(file_path: str, bank_name: str) -> dict:
    path = Path(file_path)
    if not path.exists():
        raise FileNotFoundError(f"File not found: {file_path}")

    ext = path.suffix.lower()
    if ext == ".csv":
        rows = _read_csv(path)
    elif ext in (".xlsx", ".xls"):
        rows = _read_excel(path)
    else:
        raise ValueError(f"Unsupported file format: {ext}")

    if not rows:
        return {"imported": 0, "skipped": 0}

    managers = []
    skipped = 0
    for row in rows:
        contacts = _extract_contacts_from_row(row)
        if not contacts:
            skipped += 1
            continue
        for contact in contacts:
            contact["bank_name"] = bank_name
            managers.append(contact)

    if not managers:
        return {"imported": 0, "skipped": skipped}

    conn = getconn()
    try:
        _insert_managers(conn, managers)
    finally:
        putconn(conn)

    return {"imported": len(managers), "skipped": skipped}

File: lib/importManagers.ts
/**
 * Bank manager importer: reads Excel/CSV files and inserts manager records
 * into the PostgreSQL `bank_managers` table.
 *
 * Ported from Python import_managers.py → TypeScript.
 */

/**
 * Bank manager importer: reads Excel/CSV files and inserts manager records
 * into the PostgreSQL `bank_managers` table.
 *
 * Ported from Python import_managers.py → TypeScript.
 * Internal usage: bank-managers file upload API route.
 * Depends on: lib/db (PostgreSQL pool), xlsx (npm package)
 */

import pool from "./db";

function normalize(value: any): string | null {
  if (value == null) return null;
  return String(value).trim();
}

const PHONE_RE = /^\+?\d[\d\s\-]{7,15}$/;
const NAME_RE = /^[A-Za-z\s\.]{2,}$/;

function looksLikePhone(value: string | null): boolean {
  if (!value) return false;
  return PHONE_RE.test(value);
}

function looksLikeName(value: string | null): boolean {
  if (!value) return false;
  return NAME_RE.test(value);
}

function firstNonNull(row: Record<string, any>, aliases: string[]): string | null {
  const lowered = new Map<string, string>();
  for (const k of Object.keys(row)) {
    lowered.set(k.toLowerCase(), k);
  }
  for (const alias of aliases) {
    const key = lowered.get(alias.toLowerCase());
    if (key) {
      const value = normalize(row[key]);
      if (value) return value;
    }
  }
  return null;
}

function extractContactsFromRow(row: Record<string, any>): Array<{
  name: string | null;
  phone: string | null;
  location: string | null;
  role: string | null;
}> {
  const contacts: any[] = [];

  const rsmName = firstNonNull(row, ["rsm name", "rsm"]);
  const rsmPhone = firstNonNull(row, [
    "ph no",
    "rsm\nphone number",
    "phone",
    "mobile",
    "mobile_no",
    "contact",
    "phone_number",
    "mobile_number",
    "sm phone number",
  ]);
  const rsmLocation = firstNonNull(row, ["location", "city", "branch", "branch_name", "place"]);
  if (rsmName || rsmPhone) {
    contacts.push({
      name: looksLikeName(rsmName) ? rsmName : looksLikeName(rsmPhone) ? rsmPhone : null,
      phone: looksLikePhone(rsmPhone) ? rsmPhone : looksLikePhone(rsmName) ? rsmName : null,
      location: rsmLocation,
      role: "RSM",
    });
  }

  const smName = firstNonNull(row, ["sm name", "sm"]);
  const smPhone = firstNonNull(row, [
    "ph no_1",
    "rsm\nphone number",
    "phone",
    "mobile",
    "mobile_no",
    "contact",
    "phone_number",
    "mobile_number",
    "sm phone number",
  ]);
  const smLocation = firstNonNull(row, ["location_1", "location", "city", "branch", "branch_name", "place"]);
  if (smName || smPhone) {
    contacts.push({
      name: looksLikeName(smName) ? smName : looksLikeName(smPhone) ? smPhone : null,
      phone: looksLikePhone(smPhone) ? smPhone : looksLikePhone(smName) ? smName : null,
      location: smLocation,
      role: "SM",
    });
  }

  const coordinatorName = firstNonNull(row, ["coordinator", "co-ordinator", "name"]);
  const coordinatorPhone = firstNonNull(row, [
    "mobile",
    "phone",
    "mobile_no",
    "contact",
    "phone_number",
    "mobile_number",
    "rsm\nphone number",
    "sm phone number",
  ]);
  const coordinatorLocation = firstNonNull(row, [
    "location_2",
    "location_1",
    "location",
    "city",
    "branch",
    "branch_name",
    "place",
  ]);
  if (coordinatorName || coordinatorPhone) {
    contacts.push({
      name: looksLikeName(coordinatorName) ? coordinatorName : looksLikeName(coordinatorPhone) ? coordinatorPhone : null,
      phone: looksLikePhone(coordinatorPhone) ? coordinatorPhone : looksLikePhone(coordinatorName) ? coordinatorName : null,
      location: coordinatorLocation,
      role: "Coordinator",
    });
  }

  const asmName = firstNonNull(row, ["asm/sm/dsm name", "asm", "sm", "dsm"]);
  const asmPhone = firstNonNull(row, [
    "mobile",
    "phone",
    "mobile_no",
    "contact",
    "phone_number",
    "mobile_number",
    "rsm\nphone number",
    "sm phone number",
  ]);
  const asmLocation = firstNonNull(row, [
    "location_2",
    "location_1",
    "location",
    "city",
    "branch",
    "branch_name",
    "place",
  ]);
  if (asmName || asmPhone) {
    contacts.push({
      name: looksLikeName(asmName) ? asmName : looksLikeName(asmPhone) ? asmPhone : null,
      phone: looksLikePhone(asmPhone) ? asmPhone : looksLikePhone(asmName) ? asmName : null,
      location: asmLocation,
      role: "ASM/SM/DSM",
    });
  }

  const genericName = firstNonNull(row, ["name", "manager_name", "manager", "employee_name", "contact_person"]);
  const genericPhone = firstNonNull(row, [
    "mobile",
    "phone",
    "mobile_no",
    "contact",
    "phone_number",
    "mobile_number",
    "rsm\nphone number",
    "sm phone number",
  ]);
  const genericLocation = firstNonNull(row, ["location", "city", "branch", "branch_name", "place"]);
  if (genericName || genericPhone) {
    contacts.push({
      name: looksLikeName(genericName) ? genericName : looksLikeName(genericPhone) ? genericPhone : null,
      phone: looksLikePhone(genericPhone) ? genericPhone : looksLikePhone(genericName) ? genericName : null,
      location: genericLocation,
      role: null,
    });
  }

  return contacts;
}

async function insertManagers(client: any, managers: any[]) {
  for (const m of managers) {
    await client.query(
      `INSERT INTO bank_managers (bank_name, name, email, phone, location, role, status)
       VALUES ($1, $2, $3, $4, $5, $6, $7)`,
      [
        m.bank_name,
        m.name || "Unknown",
        m.email || `unknown_${Math.random().toString(36).slice(2)}@example.com`,
        m.phone,
        m.location || "Unknown",
        m.role,
        "active",
      ]
    );
  }
}

export async function importManagersFromFile(filePath: string, bankName: string): Promise<{ imported: number; skipped: number }> {
  const fs = await import("fs");
  const path = await import("path");
  const xlsx = await import("xlsx");

  if (!fs.existsSync(filePath)) {
    throw new Error(`File not found: ${filePath}`);
  }

  const ext = path.extname(filePath).toLowerCase();
  let rows: Record<string, any>[] = [];

  if (ext === ".csv") {
    const csv = fs.readFileSync(filePath, "utf-8");
    const lines = csv.split(/\r?\n/).filter((line) => line.trim());
    if (lines.length === 0) return { imported: 0, skipped: 0 };
    const headers = lines[0].split(",").map((h) => h.trim().replace(/^"|"$/g, ""));
    for (let i = 1; i < lines.length; i++) {
      const values = lines[i].split(",").map((v) => v.trim().replace(/^"|"$/g, ""));
      const row: Record<string, any> = {};
      headers.forEach((h, idx) => {
        row[h] = values[idx];
      });
      rows.push(row);
    }
  } else if (ext === ".xlsx" || ext === ".xls") {
    const workbook = xlsx.readFile(filePath);
    const sheetName = workbook.SheetNames[0];
    const worksheet = workbook.Sheets[sheetName];
    const jsonData = xlsx.utils.sheet_to_json<Record<string, any>>(worksheet, { defval: "" });
    rows = jsonData;
  } else {
    throw new Error(`Unsupported file format: ${ext}`);
  }

  if (rows.length === 0) return { imported: 0, skipped: 0 };

  const managers: any[] = [];
  let skipped = 0;

  for (const row of rows) {
    const contacts = extractContactsFromRow(row);
    if (contacts.length === 0) {
      skipped++;
      continue;
    }
    for (const contact of contacts) {
      managers.push({
        ...contact,
        bank_name: bankName,
      });
    }
  }

  if (managers.length === 0) return { imported: 0, skipped };

  const client = await pool.connect();
  try {
    await insertManagers(client, managers);
    return { imported: managers.length, skipped };
  } finally {
    client.release();
  }
}
